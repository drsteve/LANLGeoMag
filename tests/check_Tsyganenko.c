#include <check.h>
#include "../libLanlGeoMag/Lgm/Lgm_MagModelInfo.h"

/*
 *  Regression tests for Tsyganenko model calculations
 */


Lgm_MagModelInfo    *mInfo;

void Tsyganenko_Setup(void) {
    mInfo = Lgm_InitMagInfo();
    return;
}

void Tsyganenko_TearDown(void) {
    Lgm_FreeMagInfo( mInfo );
    return;
}


START_TEST(test_Tsyganenko_01) {
    Lgm_Vector        Bexpect, Pos, Btest, Udiff;
    int               nTests, nPass, nFail, transflag, Passed=FALSE;
    int               retVal;
    double            del, Kp;
    char              buff[262], extModel[10], intModel[10];
    char              IsoDate[80];
    long long         TT2000;
    FILE              *testfile, *outfile;
    Lgm_DateTime      d;

    int makeNew = 1;

    /* read test file */
    testfile = fopen("check_Tsyganenko.expected","r");
    if (makeNew) outfile = fopen("check_Tsyganenko.got", "w");

    /* step through test cases one line at a time */
    nTests = 0;
    nPass = 0;
    nFail = 0;
    Passed = TRUE;
    while( fgets(buff,260,testfile) != NULL) {
        if (buff[0]!='#') {
            // read line
            sscanf(buff,"%s %s %s %lf %lf %lf %lf %lf %lf %lf",
		   &IsoDate[0], &extModel[0], &intModel[0], &Kp, 
                   &Pos.x, &Pos.y, &Pos.z, &Bexpect.x, &Bexpect.y, &Bexpect.z);

	    Lgm_InitMagInfoDefaults( mInfo );
            IsoTimeStringToDateTime( IsoDate, &d, mInfo->c );
            Lgm_Set_Coord_Transforms( d.Date, d.Time, mInfo->c );
	    if (!strncmp(intModel, "IGRF", 10)) {
	        Lgm_Set_Lgm_B_IGRF_InternalModel( mInfo );
	    }
	    else if (!strncmp(intModel, "CDIP", 10)) {
	        Lgm_Set_Lgm_B_cdip_InternalModel( mInfo );
	    }
	    else if (!strncmp(intModel, "EDIP", 10)) {
	        Lgm_Set_Lgm_B_edip_InternalModel( mInfo );
	    }
	    else {
	        nFail++;
	        printf("Test %d bad internal model %s\n", nTests, intModel);
		continue;
	    }
	    if (!strncmp(extModel, "T89", 10)) {
	        Lgm_Set_Lgm_B_T89(mInfo);
	    }
	    else {
	        nFail++;
	        printf("Test %d bad external model %s\n", nTests, extModel);
		continue;
	    }
	    mInfo->Kp = Kp;
	    retVal = mInfo->Bfield(&Pos, &Btest, mInfo);
	    if (retVal != 1) {
	        nFail++;
	        printf("Test %d odd return from Lgm_B_T89\n", nTests);
	        continue;
		}
            //printf("IsoDate = %s; d.Date, d.time = %ld, %lf \n", IsoDate, d.Date, d.Time);
            Udiff.x = Btest.x - Bexpect.x;
            Udiff.y = Btest.y - Bexpect.y;
            Udiff.z = Btest.z - Bexpect.z;
            del = Lgm_Magnitude(&Udiff);
            nTests++;
            if (fabs(del) <= 1.0e-5) {
                nPass++;
                printf("Test %d passed\n", nTests);
                }
            else {
                nFail++;
                printf("*****  warning : difference >= 1.0e-5 nT  *****\n");
                printf("Test %d failed (diff: %g %g %g   %g)\n", nTests, Udiff.x, Udiff.y, Udiff.z, fabs(del));
                }
            if (makeNew) fprintf(outfile, "%s %s %s %lf %lf %lf %lf %lf %lf %lf\n", IsoDate, extModel, intModel, Kp, Pos.x, Pos.y, Pos.z, Btest.x, Btest.y, Btest.z);

            }
        else {
            if (makeNew) fprintf(outfile, "%s", buff);
            }
        }
    if (nFail>0) Passed = FALSE;
    fclose(testfile);
    if (makeNew) fclose(outfile);
    printf("Result: %d tests pass; %d tests fail (Precision=1.0e-5 nT)\n", nPass, nFail);
    fflush(stdout);

    ck_assert_msg( Passed, "Tsyganenko tests failed.\n" );

}
END_TEST


Suite *Tsyganenko_suite(void) {

  Suite *s = suite_create("TSYGANENKO_TESTS");

  TCase *tc_Tsyganenko = tcase_create("Tsyganenko Models");
  tcase_add_checked_fixture(tc_Tsyganenko, Tsyganenko_Setup, Tsyganenko_TearDown);

  tcase_add_test(tc_Tsyganenko, test_Tsyganenko_01);

  suite_add_tcase(s, tc_Tsyganenko);

  return s;

}

int main(void) {

    int      number_failed;
    Suite   *s  = Tsyganenko_suite();
    SRunner *sr = srunner_create(s);

    printf("\n\n");
    srunner_run_all(sr, CK_NORMAL);
    number_failed = srunner_ntests_failed(sr);
    srunner_free(sr);

    return (number_failed == 0) ? EXIT_SUCCESS : EXIT_FAILURE;

}
