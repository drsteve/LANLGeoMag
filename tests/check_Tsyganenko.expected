#{
#    "CreationDate": "2019-04-09",
#    "LanlGeoMagRevision": "122e2a8b179210d7d5d7e7d332bb03fa24bc9a10",
#    "SunVector": "LOW_ACCURACY",
#    "DateTime": {
#        "DIMENSION": [1], 
#        "START_COLUMN": 0
#    }, 
#    "ExtModel": {
#        "DIMENSION": [1], 
#        "START_COLUMN": 1
#    }, 
#    "Kp": {
#        "DIMENSION": [1], 
#        "START_COLUMN": 2
#    }, 
#    "Result": {
#        "DIMENSION": [3], 
#        "START_COLUMN": 6, 
#        "units": "nT"
#    }, 
#    "TestPosition": {
#        "DIMENSION": [3], 
#        "START_COLUMN": 3, 
#        "units": "km"
#    }
#}
2005-08-31T09:00:00.000000 T89 IGRF 0 -6.6 0 0 -18.927076539382270 -1.857488346594122 80.052089573497625
2005-08-31T09:00:00.000000 T89 IGRF 1 -6.6 0 0 -20.784132182565948 -1.857488346594122 74.280950924408401
2005-08-31T09:00:00.000000 T89 IGRF 2 -6.6 0 0 -22.513829891596075 -1.857488346594122 70.183320782457827
2005-08-31T09:00:00.000000 T89 IGRF 3 -6.6 0 0 -26.367918458616543 -1.857488346594122 64.303840068519463
2009-01-01T00:00:00.000000 T89 IGRF 2 3.0 4.0 3.0 -123.994265 -73.755273 77.986302
2009-01-01T00:00:00.000000 T89 CDIP 2 3.0 4.0 3.0 -123.603699 -75.262705 77.740849
2009-01-01T00:00:00.000000 T89 EDIP 2 3.0 4.0 3.0 -122.127613 -75.432215 79.463218
